sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -31,6 +31,84 @@
+    """
+    If at least 2 tensors are panned in, the shards of all unpinned tensors are transfered to be on the same devices as those of the pinned tensors.
+    """
+    def tensor_w_shards_moved(tensor: ShardedTensor, new_devices: Tuple[int]) -> ShardedTensor:
+        """
+        Create a copy of the passed in tensor, but with shards now placed on the new devices.
+        """
+        new_shards = tuple(
+            (
+                transfer_to_logical_device(shard, new_devices[j])
+                if new_devices[j] != tensor.devices[j]
+                else barrier_on_logical_device(shard, new_devices[j])
+            )
+            for j, shard in enumerate(tensor.shards)
+        )
+        return tensor.clone(ts=new_shards, devices=new_devices)
+
+    if len(tensors) <= 1:
+        return tensors
+    assert all(isinstance(tensor, ShardedTensor) for tensor in tensors)
+    
+    # Check if all tensors are on the same devices.
+    all_on_same_devices = True
+    for tensor in tensors[1:]:
+        if any(d0 != d for d0, d in zip(tensors[0].devices, tensor.devices)):
+            all_on_same_devices = False
+            break
+    if all_on_same_devices:
+        return tensors
+
+    pinned_tensors = [tensor for tensor in tensors if tensor.pinned]
+    if len(pinned_tensors) == 0:
+        raise ValueError("Tensors are on different devices, but none are pinned. Don't know which devices to transfer to.")
+    
+    pinned_devices = pinned_tensors[0].devices
+    for pinned_tensor in pinned_tensors[1:]:
+        if any(d0 != d for d0, d in zip(pinned_devices, pinned_tensor.devices)):
+            raise ValueError("All pinned tensors must be on the same devices.")
+
+    # Move all non-pinned tensors to the same devices as the pinned ones.
+    new_tensors =  tuple(
+        (
+            tensor
+            if tensor.pinned
+            else tensor_w_shards_moved(tensor, pinned_devices)
+        )
+        for tensor in tensors
+    )
+    return new_tensors
+
+
+def override_w_tranfer_n_pin(operation, *override_args, **override_kwargs):
+    """
+    Wrapper for operation.override that transfers tensors to the same devices if needed.
+    Also, for uniary operations, it will pin the output tensor if the input tensor is pinned.
+    """
+    def decorator(f):
+        @operation.override(*override_args, **override_kwargs)
+        def wrapper(*args: Tuple, **kwargs: Dict[str, Any]):
+            t_i = [i for i, arg in enumerate(args) if isinstance(arg, ShardedTensor)]
+            t_vals = transfer_if_needed(*(args[i] for i in t_i))
+
+            args = list(args)
+            for i, t in zip(t_i, t_vals):
+                args[i] = t
+            args = tuple(args)
+            res = f(*args, **kwargs)
+            if (
+                len(t_i) == 1
+                and t_i[0].pinned
+                and isinstance(res, ShardedTensor)
+            ):
+                res.pinned = True
```

Comment:
We should set device information as well. The idea is the override should set all the additional device information based on the invocation.